lib/src/Uncertainty/Distribution/VonMisesFactory.cxx

```diff
@@ -0,0 +1,168 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/VonMisesFactory.hxx"
+#include "openturns/MethodBoundEvaluation.hxx"
+#include "openturns/SpecFunc.hxx"
+#include "openturns/Brent.hxx"
+#include "openturns/Point.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(VonMisesFactory)
+
+static const Factory<VonMisesFactory> Factory_VonMisesFactory;
+
+/* Default constructor */
+VonMisesFactory::VonMisesFactory(const Bool circular)
+  : DistributionFactoryImplementation()
+  , circular_(circular)
+{
+  // Nothing to do
+}
+
+/* Virtual constructor */
+VonMisesFactory * VonMisesFactory::clone() const
+{
+  return new VonMisesFactory(*this);
+}
+
+struct VonMisesFactoryParameterConstraint
+{
+  /** Constructor from a sample and a derivative factor estimate */
+  VonMisesFactoryParameterConstraint(const Scalar logMeanCos)
+    : logMeanCos_(logMeanCos)
+  {
+    // Nothing to do
+  };
+
+  Point computeConstraint(const Point & parameter) const
+  {
+    const Scalar kappa = parameter[0];
+    if (!(kappa > 0.0)) throw InvalidArgumentException(HERE) << "Error: the kappa parameter must be positive.";
+    return Point(1, SpecFunc::DeltaLogBesselI10(kappa) - logMeanCos_);
+  }
+
+  Scalar logMeanCos_;
+};
+
+
+/* Here is the interface that all derived class must implement */
+
+Distribution VonMisesFactory::build(const Sample & sample) const
+{
+  return buildAsVonMises(sample).clone();
+}
+
+Distribution VonMisesFactory::build(const Point & parameters) const
+{
+  return buildAsVonMises(parameters).clone();
+}
+
+Distribution VonMisesFactory::build() const
+{
+  return buildAsVonMises().clone();
+}
+
+VonMises VonMisesFactory::buildAsVonMises(const Sample & sample) const
+{
+  const UnsignedInteger size = sample.getSize();
+  if (size == 0) throw InvalidArgumentException(HERE) << "Error: cannot build a VonMises distribution from an empty sample";
+  if (sample.getDimension() != 1) throw InvalidArgumentException(HERE) << "Error: can build a VonMises distribution only from a sample of dimension 1, here dimension=" << sample.getDimension();
+  if (!(sample.getMin()[0] != sample.getMax()[0])) throw InvalidArgumentException(HERE) << "Error: cannot build a VonMises distribution based on a constant sample";
+  // In the case of circular data
+  Scalar mu = 0.0;
+  Scalar ratio = 0.0;
+  if (circular_)
+    {
+      Scalar sumCosXk = 0.0;
+      Scalar sumSinXk = 0.0;
+      for (UnsignedInteger i = 0; i < size; ++i)
+        {
+          sumCosXk += std::cos(sample(i, 0));
+          sumSinXk += std::sin(sample(i, 0));
+        }
+      sumCosXk /= size;
+      sumSinXk /= size;
+      mu = std::atan2(sumSinXk, sumCosXk);
```

Comment:
@josephmure  See https://en.wikipedia.org/wiki/Atan2:
```
The C function atan2, and most other computer implementations, are designed to reduce the effort of transforming 
cartesian to polar coordinates and so always define atan2(0, 0). On implementations without signed zero, or when given 
positive zero arguments, it is normally defined as 0. It will always return a value in the range [−π, π] rather than raising 
an error or returning a NaN (Not a Number).
```

Reply:
OK. Great!